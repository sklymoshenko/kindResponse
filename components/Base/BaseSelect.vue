<template>
  <div>
    <button
      id="dropdown-button"
      type="button"
      aria-haspopup="listbox"
      aria-expanded="true"
      aria-labelledby="listbox-label"
      class="relative w-full border border-primary rounded-md pl-3 pr-10 py-2 text-left cursor-default focus:outline-none focus:ring-1 focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm excluded"
      @click="openDropDown($event)"
    >
      <span class="flex items-center">
        <span class="ml-3 block truncate">
          {{ selected.name }}
        </span>
      </span>
      <span
        class="ml-3 absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
      >
        <!-- Heroicon name: selector -->
        <svg
          class="h-5 w-5 text-gray-400"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 20 20"
          fill="currentColor"
          aria-hidden="true"
        >
          <path
            fill-rule="evenodd"
            d="M10 3a1 1 0 01.707.293l3 3a1 1 0 01-1.414 1.414L10 5.414 7.707 7.707a1 1 0 01-1.414-1.414l3-3A1 1 0 0110 3zm-3.707 9.293a1 1 0 011.414 0L10 14.586l2.293-2.293a1 1 0 011.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
            clip-rule="evenodd"
          />
        </svg>
      </span>
    </button>

    <!--
      Select popover, show/hide based on select state.
    -->
    <div
      v-blur-closing="{
        handler: 'blurHandler'
      }"
      :class="showDropdown ? 'block' : 'hidden'"
      class="absolute mt-1 w-full rounded-md bg-white shadow-lg z-10"
    >
      <ul
        tabindex="-1"
        role="listbox"
        aria-labelledby="listbox-label"
        aria-activedescendant="listbox-item-3"
        class="max-h-56 rounded-md py-1 text-base ring-1 ring-black ring-opacity-5 overflow-auto focus:outline-none sm:text-sm h-48"
      >
        <!--
          Select option, manage highlight styles based on mouseenter/mouseleave and keyboard navigation.

          Highlighted: "text-white bg-indigo-600", Not Highlighted: "text-gray-900"
        -->
        <li
          v-for="(option, i) in options"
          :id="`listbox-item-${i}`"
          :key="option.name"
          role="option"
          class="text-gray-900 cursor-default select-none relative py-2 pl-3 pr-9 hover:bg-secondary hover:text-white"
          @click="optionSelected(option)"
        >
          <div class="flex items-center">
            <!-- Selected: "font-semibold", Not Selected: "font-normal" -->
            <span class="ml-3 block font-normal truncate">
              {{ option.name }}
            </span>
          </div>

          <!--
            Checkmark, only display for selected option.
          -->
          <span
            v-if="option.selected"
            class="absolute inset-y-0 right-0 flex items-center pr-4"
          >
            <!-- Heroicon name: heart -->
            <svg
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 20 20"
              fill="currentColor"
              class="icon-test stroke-current text-danger w-8"
            >
              <path
                fill-rule="evenodd"
                d="M3.172 5.172a4 4 0 015.656 0L10 6.343l1.172-1.171a4 4 0 115.656 5.656L10 17.657l-6.828-6.829a4 4 0 010-5.656z"
                clip-rule="evenodd"
              />
            </svg>
          </span>
        </li>
      </ul>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Emit } from "nuxt-property-decorator";
import { Category } from "@/interfaces/BaseInterfaces.ts";
@Component
export default class BaseSelect extends Vue {
  @Prop({ type: Array, default: () => [{ name: "No options" }] })
  options!: Category[];

  showDropdown: boolean = false;
  isOptionHovered: boolean = false;
  openDropDown(): void {
    this.showDropdown = !this.showDropdown;
  }

  blurHandler(): void {
    this.showDropdown = false;
  }

  @Emit("option:select")
  optionSelected(option: Category): Category {
    this.showDropdown = false;
    return option;
  }

  get selected(): Category {
    const selected = this.options.find(({ selected }) => selected);
    const defaultOptions = { name: "Nothing...", url: "", fields: [] };
    return selected || defaultOptions;
  }
}
</script>

<style lang="scss" scoped></style>
