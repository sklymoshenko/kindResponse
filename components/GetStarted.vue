<template>
  <div class="started-test flex mx-auto bg-shade h-64 w-4/5 z-10 origin-center" />
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "GetStarted"
});
</script>

<style></style>
