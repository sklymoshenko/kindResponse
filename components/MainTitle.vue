<template>
  <span class="flex flex-col justify-center items-center mt-24 md:mb-8">
    <span class="flex items-end">
      <p class="title-test font-titles font-bold capitalize sm:text-lg md:text-2xl mr-1" :title="text">
        {{ text }}
      </p>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 20 20"
        fill="currentColor"
        class="icon-test stroke-current text-danger w-8"
      >
        <path
          fill-rule="evenodd"
          d="M3.172 5.172a4 4 0 015.656 0L10 6.343l1.172-1.171a4 4 0 115.656 5.656L10 17.657l-6.828-6.829a4 4 0 010-5.656z"
          clip-rule="evenodd"
        />
      </svg>
    </span>
  </span>
</template>

<script lang="ts">
import { Component, Vue } from "nuxt-property-decorator";
@Component
export default class MainTitle extends Vue {
  text: string = "Kind responses for everybody";
};
</script>

<style></style>
