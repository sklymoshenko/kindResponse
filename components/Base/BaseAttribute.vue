<template>
  <div class="rounded-md w-full my-2 relative">
    <label id="listbox-label" class="block text-sm font-medium text-gray-700 text-left pl-2">
      {{ label }}
    </label>
    <component
      :is="component"
      :id="label"
      v-bind="$props"
      :class="`focus:outline-none focus:border-secondary transition-colors duration-500 block w-full sm:text-lg rounded-sm font-primary ${textInputClasses}`"
      :placeholder="placeholder"
      autocomplete="chrome-off"
      @option:select="optionSelected"
      @input="handleInputChange"
    >
      <option value="" disabled selected>
        {{ label }}
      </option>
    </component>
  </div>
</template>

<script lang="ts">
import {
  Component,
  Prop,
  Emit,
  Vue
} from "nuxt-property-decorator";

import BaseSelect from "@/components/Base/BaseSelect.vue";
import { ComponentsMap, Category } from "@/interfaces/BaseInterfaces.ts";

@Component({
  components: {
    BaseSelect
  }
})
export default class BaseAttribute extends Vue {
  @Prop({ type: String, default: "text", required: true }) type!: string;
  @Prop({ type: String, default: "" }) placeholder!: string;
  @Prop({ type: String, default: "Lbl" }) label!: string;
  @Prop({ type: Array, default: () => [{ name: "No options" }] })
  options!: Category[];

  componentsMap: ComponentsMap = {
    text: "input",
    select: "BaseSelect"
  };

  @Emit("option:select")
  optionSelected(option: Category): Category {
    return option;
  }

  @Emit("input:changed")
  handleInputChange(e: { target: HTMLInputElement }): string {
    const input = e.target.value || "Mike";
    return input;
  }

  get component(): string {
    const type: string = this.type;
    return this.componentsMap[type as keyof ComponentsMap] || "input";
  };

  get textInputClasses(): string {
    const isTextInput: boolean = this.type === "text";
    return isTextInput ? "border-b border-solid border-primary pl-2" : "";
  }
};
</script>

<style></style>
