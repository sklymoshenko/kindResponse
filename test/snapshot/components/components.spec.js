import { mount, createLocalVue } from "@vue/test-utils";

import BlurClosing from "@/directives/blurClosing.ts";

import GetStarted from "@/components/GetStarted.vue";
import MainTitle from "@/components/MainTitle.vue";
import BaseAttribute from "@/components/Base/BaseAttribute.vue";
import BaseSelect from "@/components/Base/BaseSelect.vue";

const fetchedOptions = [{ name: "test", selected: false }, { name: "test2", selected: false }];
const localVue = createLocalVue();

localVue.directive("blur-closing", BlurClosing);

describe("Snapshot tests", () => {
  describe("Get Started", () => {
    fetch.mockResponse(JSON.stringify(fetchedOptions));
    const wrapper = mount(GetStarted, { localVue });
    test("Should create a vue instance", async() => {
      await wrapper.vm.fetchCategories();
      expect(wrapper.vm).toBeTruthy();
    });
    test("Renders correctly", async() => {
      await wrapper.vm.fetchCategories();
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Shows selected category", async() => {
      await wrapper.vm.fetchCategories();
      wrapper.vm.selectCategory(fetchedOptions);
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Emitted 'to' change", async() => {
      await wrapper.vm.handleToChange();
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Emitted 'from' change", async() => {
      await wrapper.vm.handleFromChange();
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe("Main Title", () => {
    const wrapper = mount(MainTitle, { localVue });
    test("Should create a vue instance", () => {
      expect(wrapper.vm).toBeTruthy();
    });
    test("Renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe("Base components", () => {
    describe("Attribute", () => {
      const wrapper = mount(BaseAttribute, {
        propsData: { type: "text" },
        directives: { "blur-closing": () => jest.fn() }
      });
      test("Should create a a vue instance", () => {
        expect(wrapper.vm).toBeTruthy();
      });
      test("Renders correctly", () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Renders correctly with another type", async() => {
        await wrapper.setProps({ type: "select" });
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Renders correctly with default type", async() => {
        await wrapper.setProps({ type: "" });
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Select options", async() => {
        await wrapper.vm.optionSelected();
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Input changed", async() => {
        const input = wrapper.find("input#Lbl");
        await input.setValue("s");
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Empty input", async() => {
        const input = wrapper.find("input#Lbl");
        await input.setValue("");
        expect(wrapper.element).toMatchSnapshot();
      });
    });
    describe("Select", () => {
      const wrapper = mount(BaseSelect, { localVue });
      test("Should create a a vue instance", () => {
        expect(wrapper.vm).toBeTruthy();
      });
      test("Renders correctly", () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      test("openDropDown => toggle showDropDown", async() => {
        const button = wrapper.find("button#dropdown-button");
        await button.trigger("click");
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Close dropdown on blur", () => {
        wrapper.vm.blurHandler();
        expect(wrapper.element).toMatchSnapshot();
      });
      test("Click on option change selected", async() => {
        const option = wrapper.find("li#listbox-item-0");
        await option.trigger("click");
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
  describe("Chat", () => {
    const wrapper = mount(MainTitle, { localVue });
    test("Should create a vue instance", () => {
      expect(wrapper.vm).toBeTruthy();
    });
    test("Renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
