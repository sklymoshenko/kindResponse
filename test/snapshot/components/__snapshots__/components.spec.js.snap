// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Get Started Renders correctly 1`] = `
<div
  class="started-test flex mx-auto bg-shade h-64 w-4/5 z-10 origin-center"
/>
`;

exports[`Snapshot tests Main Title Renders correctly 1`] = `
<span
  class="flex flex-col justify-center items-center mt-12"
>
  <span
    class="flex items-end"
  >
    <p
      class="title-test font-titles font-bold capitalize sm:text-lg md:text-2xl mr-1"
    >
      
      Kind responses for everybody
    
    </p>
     
    <svg
      class="icon-test stroke-current text-danger w-8"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M3.172 5.172a4 4 0 015.656 0L10 6.343l1.172-1.171a4 4 0 115.656 5.656L10 17.657l-6.828-6.829a4 4 0 010-5.656z"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</span>
`;
